### ISSUE TITLE
JSONLogger doesn't produce any output if data map contains Valuer interface

### ISSUE THREAD

**Issue by pazderak** [go-kit/kit#713]
**Steps to reproduce:**
If Valuer is added as value to the logger (via log.With) and it is on the position bigger then a length of values to log (with logger.Log), then Valuer is not evaluated and JSON marshaler fails with unsupported type error. This error is hidden if we don't track output from logger.Log method.

```go
package main

import (
	"os"

	"github.com/go-kit/kit/log"
	"github.com/go-kit/kit/log/level"
)

func main() {
	logger := log.NewJSONLogger(log.NewSyncWriter(os.Stdout))
	// creating logger with four items, fourth one is a Valuer
	logger = log.With(logger,
		"@by", "",
		"@env", "",
		"data_version", 2,
		"application", log.Valuer(func() interface{} { "application" }),
	)
	// logging three items
	logger.Log(
		level.Key(), level.InfoValue(),
		"@timestamp", log.DefaultTimestampUTC,
		"description", "running",
	)
}
```
**Expected behavior:**
Log record is created properly

**Comment by ChrisHines** [comment]
I copied your example above to reproduce the problem. It turns out it is not the `log.Valuer` passed to `log.With` that is causing the problem.

The source of the problem is the `log.DefaultTimestampUTC` passed directly to the `logger.Log` call. `log.DefaultTimestampUTC` is created with `log.TimestampFormat` which returns a `log.Valuer`. It is intended for use with `log.With` and `log.WithPrefix`. If you want to pass a new timestamp as a value to `logger.Log` you can just pass `time.Now().UTC()` directly.

I hope that helps you understand what's going on.

**Comment by pazderak** [comment]
Thank you for the explanation. So I cannot use Valuer interface as parameter in the Log message, correct?
If yes, it also explains the logic in https://github.com/go-kit/kit/blob/828ac175f5d2a008e9eea3a36f474f19a774a3a3/log/log.go#L111.

Thank you again.

**Comment by ChrisHines** [comment]
That's right. Valuers should only be stashed in a contextual logger created with `log.With` or `log.WithPrefix`. That's how this package generates a new computed value for things like timestamps and source code location information each time a log event is produced without having to repeat those key/value pairs at each place the Log method is called. Although there are technical reasons that make resolving a Valuer in the Log arguments difficult, it is also unnecessary. By nature any arguments you pass to Log will be freshly computed each time the application evaluates that line of code, so wrapping them in a Valuer doesn't add any functionality at that point.

A Valuer *is* needed to defer evaluation of dynamic values stored in a contextual logger, however. Study the output of this example to see these effects in action.

```go
package main

import (
	"os"
	"time"

	"github.com/go-kit/kit/log"
)

func main() {
	start := time.Now()
	logger := log.NewLogfmtLogger(log.NewSyncWriter(os.Stdout))
	logger = log.With(logger,
		"stuck", time.Since(start),
		"valuer", log.Valuer(func() interface{} { return time.Since(start) }),
	)
	for i := 0; i < 5; i++ {
		time.Sleep(50 * time.Millisecond)
		logger.Log("log", time.Since(start))
	}
}

```
Running the above program just now produced this output:
```
stuck=13.61µs valuer=52.910514ms log=52.869637ms
stuck=13.61µs valuer=103.925416ms log=103.910036ms
stuck=13.61µs valuer=157.998311ms log=157.985381ms
stuck=13.61µs valuer=210.240348ms log=210.231469ms
stuck=13.61µs valuer=260.581752ms log=260.569589ms
```

**Comment by ChrisHines** [comment]
@pazderak Were you able to resolve your issue? Can we close this issue?

**Comment by pazderak** [comment]
Yes, it works now. Thank you again.
